swift/Arrow/Sources/Arrow/ArrowCImporter.swift

```diff
@@ -0,0 +1,144 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+import ArrowC
+
+public class ArrowCImporter {
+    private func appendToBuffer(
+        _ cBuffer: UnsafeRawPointer?,
+        arrowBuffers: inout [ArrowBuffer],
+        byteCount: Int,
+        length: UInt,
+        nullCount: UInt? = nil) {
+        if cBuffer == nil {
+            arrowBuffers.append(ArrowBuffer.createEmptyBuffer())
+            return
+        }
+
+        let pointer = UnsafeMutableRawPointer(mutating: cBuffer)!
+        arrowBuffers.append(
+            ArrowBuffer(length: length, capacity: UInt(byteCount), rawPointer: pointer, isMemoryOwner: false))
+    }
+
+    public init() {}
+
+    public func importType(_ cArrow: String, name: String = "") ->
+        Result<ArrowField, ArrowError> {
+        do {
+            let type = try ArrowType.fromCDataFormatId(cArrow)
+            return .success(ArrowField(name, type: ArrowType(type.info), isNullable: true))
+        } catch {
+            return .failure(.invalid("\(error)"))
+        }
+    }
+
+    public func importField(_ cSchema: ArrowC.ArrowSchema) ->
+        Result<ArrowField, ArrowError> {
+        if cSchema.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cSchema.dictionary != nil {
+            return .failure(.invalid("Dictinoary types currently not supported"))
+        }
+
+        switch importType(
+            String(cString: cSchema.format), name: String(cString: cSchema.name)) {
+        case .success(let field):
+            release(cSchema)
+            return .success(field)
+        case .failure(let err):
+            return .failure(err)
+        }
+    }
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowType: ArrowType,
+        isNullable: Bool = false) -> Result<ArrowArrayHolder, ArrowError> {
+        let arrowField = ArrowField("", type: arrowType, isNullable: isNullable)
+        return importArray(cArray, arrowField: arrowField)
+    }
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowField: ArrowField) -> Result<ArrowArrayHolder, ArrowError> {
```

Comment:
This one does not seem to release the C array on error, does it?

Reply:
It does not. I had chatted with @zeroshade about the golang behavior and it doesn't auto release the data but does release the schema.  I mirrored the behavior in this impl.

Reply:
Well, you can of course design those APIs like this, but just because Arrow Go has a bizarre behavior doesn't make it a good idea to replicate it :-)

For example, Arrow C++ has a consistent behavior where it always releases arrays and schemas that it fails importing.

Reply:
Gotcha, I didn't look into the C++ behavior for this. I will update to release the array and schemas on failure.  Thank you!